### ISSUE TITLE
Lazy render for performance

### ISSUE THREAD

**Issue by unbug** [google/code-prettify#413]
In my case,the codes are very large,prettify freezes the browser every time.
So the idea is only render the hightlight style when the codes scroll into the viewport.
Work as CodeMirror's ```maxHighlightLength``` option,this will gets a great performance on both desktop and mobile broswer.
Is there a option like this or a work around for me to try it out?

**Comment by mikesamuel** [comment]
How would this interact with printing?  Do events get fired when a page is printed to make the entire viewport appear visible?

**Comment by unbug** [comment]
Yes,it does,if the source codes very long,then the highlight DOM will very large,the browser get freeze too.

**Comment by apotema** [comment]
\+ 1 
I am having performance problems as well.

**Comment by sim642** [comment]
This issue seems to be affecting [weechat](https://github.com/weechat/weechat) as well as the result of https://github.com/weechat/weechat/commit/e092bd19d342631c6b9ed0e721402264ed445701. The documentation page at https://weechat.org/files/doc/devel/weechat_plugin_api.en.html takes significant time to load.

**Comment by iy2chang** [comment]
I have the same issues as well, I have a code set roughly about 13,000 lines of code.

**Comment by mikesamuel** [comment]
I think the best way to address this would be to do the bulk of the work via web workers.
http://caniuse.com/#feat=webworkers

**Comment by apotema** [comment]
Some loading callback would also be nice.

**Comment by mikesamuel** [comment]
@apotema What's wrong with [`opt_whenDone`](https://github.com/google/code-prettify/blob/master/src/prettify.js#L1525)?

**Comment by apotema** [comment]
Nothing @mikesamuel. I forgot about this parameter. My bad.

**Comment by antoniotajuelo** [comment]
Even if the opt_whenDone parameter is properly used, parsing big chunks of code can freeze the browser, leading into unexpected results. At the time, I am switching into a server-side solution for syntax highlighting, as I really need to parse big chunks, and I cannot figure at the moment how to get it done with code-prettify.

**Comment by apotema** [comment]
Which server-side solution are you using @antoniotajuelo ?
